File: projects/algorithms-and-leetcode/leetcode/150_top_problems/python/merge_sorted_array.py
"""
You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.
"""

from typing import List

class Solution:
    def merge(self,nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # This is a two pointer approach
        nums1[:m] = nums2
        nums1.sort()
        return nums1

File: projects/algorithms-and-leetcode/leetcode/150_top_problems/c++/merge_sorted_array.cpp
// You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.

#include <iostream>
#include <vector>
using namespace std;


class SOlution {
    public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m - 1; // Pointer for nums1
        int j = n - 1; // Pointer for nums2
        int k = m + n - 1; // Pointer for the merged array
        while (j >= 0) {
            if (i >= 0 && nums1[i] > nums2[j]) {
                nums1[k--] = nums1[i--]; // Place larger element at the end of nums1
            } else {
                nums1[k--] = nums2[j--]; // Place element from nums2
            }               
        }
    }
};
